src/box/txn.c

```diff
@@ -850,7 +850,17 @@ txn_journal_entry_new(struct txn *txn)
 
 	struct xrow_header **remote_row = req->rows;
 	struct xrow_header **local_row = req->rows + txn->n_applier_rows;
-	bool is_sync = false;
+	bool is_sync = txn_has_flag(txn, TXN_WAIT_ACK);
+
+	/*
+	 * Can't commit a fully local transaction synchronously.
+	 */
+	if (is_sync && txn_is_fully_local(txn)) {
+		diag_set(IllegalParams, "An attempt to commit a fully local "
+			 "transaction synchronously");
+		return NULL;
+	}
+
```

Comment:
This better be moved into the existing `if` below, inside `if (!txn_has_flag(txn, TXN_FORCE_ASYNC) && !is_fully_nop) {` branch. Then we won't have this additional check of `is_sync` for regular transactions.